services/user/Tokens/service/src/lib.rs

```diff
@@ -0,0 +1,288 @@
+        Token::get_by_symbol(symbol).map(|token| token.into())
+    }
+
+    #[action]
+    fn map_symbol(token_id: TID, symbol: AccountNumber) {
+        let mut token = Token::get_assert(token_id);
+        let sender = get_sender();
+        token.check_is_owner(sender);
+        token.map_symbol(symbol);
+
+        let symbol_owner_nft = Symbol::call_from(Wrapper::SERVICE)
+            .getSymbol(symbol)
+            .ownerNft;
+
+        Nfts::call_from(Wrapper::SERVICE).debit(
+            symbol_owner_nft,
+            format!(
+                "Mapping symbol {} to token {}",
+                symbol.to_string(),
+                token_id
+            ),
+        );
+        Nfts::call_from(Wrapper::SERVICE).burn(symbol_owner_nft);
+
+        Wrapper::emit()
+            .history()
+            .symbol_mapped(token_id, sender, symbol);
+    }
+
+    #[action]
+    #[allow(non_snake_case)]
+    fn getUserConf(account: AccountNumber) -> Holder {
+        Holder::get_or_new(account)
+    }
+
+    #[action]
+    #[allow(non_snake_case)]
+    fn setUserConf(index: u8, enabled: bool) {
+        Holder::get_or_new(get_sender()).set_settings(index, enabled);
+    }
+
+    #[action]
+    #[allow(non_snake_case)]
+    fn getTokHoldr(account: AccountNumber, token_id: TID) -> TokenHolder {
+        TokenHolder::get_or_new(account, token_id)
+    }
+
+    #[action]
+    #[allow(non_snake_case)]
+    fn setTokHoldr(token_id: TID, index: u8, enabled: bool) {
+        TokenHolder::get_or_new(get_sender(), token_id).set_settings(index, enabled);
+    }
+
+    #[action]
+    #[allow(non_snake_case)]
+    fn setTokenConf(token_id: TID, index: u8, enabled: bool) {
+        let mut token = Token::get_assert(token_id);
+        token.check_is_owner(get_sender());
+        token.set_settings(index, enabled);
+    }
+
+    #[action]
+    fn open(token_id: TID) {
+        Balance::add(get_sender(), token_id);
+    }
+
+    #[action]
+    #[allow(non_snake_case)]
+    fn getBalance(token_id: TID, user: AccountNumber) -> Balance {
+        Balance::get_or_new(user, token_id)
+    }
+
+    #[action]
+    #[allow(non_snake_case)]
+    fn getSharedBal(
+        creditor: AccountNumber,
+        debitor: AccountNumber,
+        token_id: TID,
+    ) -> SharedBalance {
+        SharedBalance::get_or_new(creditor, debitor, token_id)
+    }
+
+    #[action]
+    fn recall(token_id: TID, from: AccountNumber, amount: Quantity, memo: String) {
+        check(amount.value > 0, "must be greater than 0");
+        let sender = get_sender();
+        let mut token = Token::get_assert(token_id);
+        let token_settings = token.settings();
+
+        check(sender != from, "cannot recall from self, use burn instead");
+        check(!token_settings.is_unrecallable(), "token is not recallable");
+
+        token.check_is_owner(sender);
+        token.burn(amount, from);
+
+        Wrapper::emit()
+            .history()
+            .recall(token_id, amount, sender, from, memo);
```

Comment:
@swatanabe is it possible for an arbitrary event to get the blockNum in which it was written or should we be adding timestamps/blocknums to events?